topomodelx/nn/hypergraph/unigcnii_layer.py

```diff
@@ -0,0 +1,101 @@
+"""UniGCNII layer implementation."""
+import torch
+
+
+class UniGCNIILayer(torch.nn.Module):
+    r"""Implementation of the UniGCNII layer.
+
+    The later should always be used with self loops on all nodes (should this be tested?).
+
+    Parameters
+    ----------
+    in_channels : int
+        Number of input channels of the node (rank-0) features.
+    alpha : float
+        The alpha parameter determining the importance of the self-loop (\theta_2).
+    beta : float
+        The beta parameter determining the importance of the learned matrix (\theta_1).
+    """
+
+    def __init__(self, in_channels, alpha, beta):
+        super().__init__()
+
+        self.in_channels = in_channels
+        self.alpha = alpha
+        self.beta = beta
+        self.linear = torch.nn.Linear(in_channels, in_channels, bias=False)
+
+    def reset_parameters(self):
+        """Reset the parameters of the layer."""
+        self.linear.reset_parameters()
+
+    def forward(self, x_0, incidence_1, x_skip=None):
+        r"""Forward pass of the UniGCNII layer.
+
+        The forward pass consists of two messages and a skip connection with a learned update function.
+
+        1. Every hyper-edge sums up the features of its constituent edges:
+        .. math::
+            \begin{align*}
+            & 🟥 \quad m_{y \rightarrow z}^{(0 \rightarrow 1)} = (B^T_1)\_{zy} \cdot h^{t,(0)}_y \\
+            & 🟧 \quad m_z^{(0\rightarrow1)} = \sum_{y \in \mathcal{B}(z)} m_{y \rightarrow z}^{(0 \rightarrow 1)}
+            \end{align*}
+
+        2. The second message is normalized with the node and edge degrees:
+        .. math::
+            \begin{align*}
+            & 🟥 \quad m_{z \rightarrow x}^{(1 \rightarrow 0)}  = B_1 \cdot m_z^{(0 \rightarrow 1)} \\
+            & 🟧 \quad m_{x}^{(1\rightarrow0)}  = \frac{1}{\sqrt{d_x}}\sum_{z \in \mathcal{C}(x)} \frac{1}{\sqrt{d_z}}m_{z \rightarrow x}^{(1\rightarrow0)} \\
+            \end{align*}
+
+        3. The computed message is combined with skip connections and a linear transformation using hyperparameters alpha and beta:
+        .. math::
+            \begin{align*}
+            & 🟩 \quad m_x^{(0)}  = m_x^{(1 \rightarrow 0)} \\
+            & 🟦 \quad m_x^{(0)}  = ((1-\beta)I + \beta W)((1-\alpha)m_x^{(0)} + \alpha \cdot h_x^{t,(0)}) \\
+            \end{align*}
+
+        Parameters
+        ----------
+        x_0 : torch.Tensor, shape = [num_nodes, in_channels]
+            Input features of the nodes of the hypergraph.
+        incidence_1 : torch.Tensor, shape = [num_nodes, num_edges]
+            Incidence matrix of the hypergraph.
+        x_skip : torch.Tensor, shape = [num_nodes, in_channels]
+            Original node features of the hypergraph used for the skip connections.
+
+        Returns
+        -------
+        _ : torch.Tensor, shape = [num_nodes, in_channels]
+            Output features of the nodes of the hypergraph.
+
+        """
+        x_skip = x_0 if x_skip is None else x_skip
```

Comment:
The original paper introduces skip connections from the input features on the nodes $h_x^{0,(0)}$, while the [message passing equations ](https://github.com/awesome-tnns/awesome-tnns/blob/main/Hypergraphs.md), use the input to the layer as skip connection $h_x^{t,(0)}$. For now, the parameter x_skip can be used to determine the behaviour. If not set the input to the layer is used as a skip connection, otherwise the provided features.